### PR TITLE
zkSync L1 gas Price calculation oracle

### PR DESCRIPTION
**Pull request by simsonraj**
ZkSync L1 Gas price calculation by multiplying gasPerPubDataByte * L2GasPrice

### CODE REVIEW

File: core/chains/evm/gas/rollups/zkSync_l1_oracle.go [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+		return t, err
+	}
+
+	o.l1GasPriceMu.Lock()
+	defer o.l1GasPriceMu.Unlock()
+	o.l1GasPrice = priceEntry{price: assets.NewWei(price), timestamp: time.Now()}
+	return
+}
+
+// For zkSync l1_price_per_pubdata ~= gas_price_on_l2 * l2_gas_PerPubdataByte
```
Comment (matYang) [inline core/chains/evm/gas/rollups/zkSync_l1_oracle.go]: nit: explicitly specify is it per byte, like `l1_price_per_pubdata_byte`, as opposed to pubdata gas or slot

File: core/chains/evm/gas/rollups/zkSync_l1_oracle.go [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+	// Allowing double the poll period before declaring the price stale to give ample time for the refresh to process
+	if time.Since(timestamp) > o.pollPeriod*2 {
+		return l1GasPrice, fmt.Errorf("gas price is stale")
+	}
+	return
+}
+
+// Gets the L1 gas cost for the provided transaction at the specified block num
+// If block num is not provided, the value on the latest block num is used
+func (o *zkSyncL1Oracle) GetGasCost(ctx context.Context, tx *gethtypes.Transaction, blockNum *big.Int) (*assets.Wei, error) {
```
Comment (matYang) [inline core/chains/evm/gas/rollups/zkSync_l1_oracle.go:185]: Is returning 0 here safe for downstream code?

**Review by matYang — DISMISSED** [review]

File: core/chains/evm/gas/rollups/zkSync_l1_oracle.go [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+
+func NewZkSyncL1GasOracle(lggr logger.Logger, ethClient l1OracleClient) *zkSyncL1Oracle {
+	var systemContextAddress, gasPerPubdataMethod, gasPerPubdataSelector, l2GasPriceMethod, l2GasPriceSelector string
+
+	systemContextAddress = SystemContextAddress
+	gasPerPubdataMethod = SystemContext_gasPerPubdataByteMethod
+	gasPerPubdataSelector = ZksyncGasInfo_getGasPerPubdataByteL2
+
+	l2GasPriceMethod = SystemContext_gasPriceMethod
+	l2GasPriceSelector = ZksyncGasInfo_getGasPriceL2
```
Comment (amit-momin) [inline core/chains/evm/gas/rollups/zkSync_l1_oracle.go]: Couldn't we just set the fields in the struct directly?

Comment (simsonraj) [reply core/chains/evm/gas/rollups/zkSync_l1_oracle.go]: yea good point, just followed the other oracles implementation, and just realised its unnecessary

Comment (amit-momin) [reply core/chains/evm/gas/rollups/zkSync_l1_oracle.go]: Noticed that in the others too. Not sure why they were written like that tbh. Likely in need of a clean up.

Comment (simsonraj) [reply core/chains/evm/gas/rollups/zkSync_l1_oracle.go]: yeah, I cleaned up zkSync anyway, good catch

**Review by amit-momin — APPROVED** [review]

**Review by FelixFan1992 — APPROVED** [review]

**Review by matYang — APPROVED** [review]

**Pull request merged**